src/shogun/mathematics/linalg/LinalgBackendEigen.h

```diff
@@ -291,6 +310,30 @@ class LinalgBackendEigen : public LinalgBackendBase
 		return (typename SGVector<T>::EigenVectorXtMap(a)).dot(typename SGVector<T>::EigenVectorXtMap(b));
 	}
 
+	/** Eigen3 matrix elementwise product method */
+	template <typename T>
+	SGMatrix<T> element_prod_impl(const SGMatrix<T>& a, const SGMatrix<T>& b) const
+	{
+		SGMatrix<T> c(a.num_rows, a.num_cols);
+		typename SGMatrix<T>::EigenMatrixXtMap a_eig = a;
+		typename SGMatrix<T>::EigenMatrixXtMap b_eig = b;
+		typename SGMatrix<T>::EigenMatrixXtMap c_eig = c;
+
+		c_eig = a_eig.array() * b_eig.array();
+		return c;
+	}
+
+	/** Eigen3 matrix in-place elementwise product method */
+	template <typename T>
+	void element_prod_impl(SGMatrix<T>& a, SGMatrix<T>& b, SGMatrix<T>& result) const
```

Comment:
@karlnapf we had discussions in #3387 and #3391. So it seems that two have two keep two interfaces. But I am now thinking maybe I can call refactor current `Matrix func(Matrix a, Matrixb)` to:
```
alloc(result, num_rows, num_cols);
func_in_place(a, b, result);
return result
```
So the codes won't be duplicated.

Also I am thinking whether it's possible to have a wrapper method for all non-in-place methods, such as
```
Container<T> wrapper(Container<T> a, Container<T> b, lambda x)
{
    alloc(result, size);
    x(a, b, result);
    return result;
}